app/models/course/lesson_plan/item.rb

```diff
@@ -93,6 +108,39 @@ class Course::LessonPlan::Item < ApplicationRecord
            to: :default_reference_time
   before_validation :link_default_reference_time
 
+  # Returns a frozen CourseReferenceTime or CoursePersonalTime.
+  # The calling function is responsible for eager-loading both associations if calling time_for on a lot of items.
+  # TODO(#3902): Lookup user's reference timeline before defaulting to default reference timeline
+  def time_for(course_user)
+    personal_time = eager_loaded_personal_time_for(course_user)
+    reference_time = eager_loaded_reference_time_for(course_user)
+    (personal_time || reference_time).dup.freeze
+  end
+
+  def eager_loaded_personal_time_for(course_user)
+    # Do not make a separate call to DB if personal_times has already been preloaded
+    if personal_times.loaded?
+      personal_times.find { |x| x.course_user == course_user }
+    else
+      personal_times.find_by(course_personal_times: { course_user: course_user })
+    end
+  end
+
+  def eager_loaded_reference_time_for(course_user)
+    # Do not make a separate call to DB if reference_times has already been preloaded
+    reference_timeline_id = course_user.reference_timeline_id || course_user.course.default_reference_timeline.id
+    if reference_times.loaded?
+      reference_times.find { |x| x.reference_timeline_id == reference_timeline_id }
+    else
+      reference_times.find_by(course_reference_times: { reference_timeline_id: reference_timeline_id })
+    end
+  end
+
+  # Gets the existing personal time for course_user, or instantiates and returns a new one
+  def setdefault_personal_time_for(course_user)
```

Comment:
I think we should combine this with `eager_loaded_personal_time_for`, name it `personal_time_for`, and have it init the values to the reference time.

Right now, if I do `item.setdefault_personal_time_for(course_user)`, I can't actually save the item since the times haven't been set at all, which makes the function name rather misleading.

Reply:
Doing this on `eager_loaded_personal_time_for` is not an option because then we can't differentiate between an existing personalized timeline and a newly instantiated personalized timeline (which we often won't save).

Function name was taken from Python which does something similar: https://www.tutorialspoint.com/python/dictionary_setdefault.htm